File: typescript/src/resources/contacts.ts
/**
 * Contacts resource - Contact and contact list management
 */

import type { HttpClient } from '../utils/http';
import type {
  Contact,
  CreateContactRequest,
  UpdateContactRequest,
  ContactList,
  CreateContactListRequest,
  UpdateContactListRequest,
  BulkContactResult,
  ContactImportJob,
  PaginatedResponse,
} from '../types';
import type { RetryConfig } from '../utils/retry';
import { withRetry } from '../utils/retry';

export class ContactsResource {
  constructor(
    private http: HttpClient,
    private retryConfig: RetryConfig
  ) {}

  // ========== Contacts ==========

  /**
   * List contacts
   */
  public async list(options?: {
    page?: number;
    limit?: number;
    search?: string;
    listId?: string;
    tags?: string;
  }): Promise<PaginatedResponse<Contact>> {
    const params = new URLSearchParams();
    if (options?.page) params.set('page', options.page.toString());
    if (options?.limit) params.set('limit', options.limit.toString());
    if (options?.search) params.set('search', options.search);
    if (options?.listId) params.set('listId', options.listId);
    if (options?.tags) params.set('tags', options.tags);
    
    return withRetry(
      () => this.http.get(`/api/v1/contacts?${params.toString()}`),
      this.retryConfig
    );
  }

  /**
   * Create a new contact
   */
  public async create(request: CreateContactRequest): Promise<{ success: true; data: Contact }> {
    return withRetry(
      () => this.http.post('/api/v1/contacts', request),
      this.retryConfig
    );
  }

  /**
   * Delete a contact
   */
  public async delete(id: string): Promise<{ success: true }> {
    return withRetry(
      () => this.http.delete(`/api/v1/contacts/${id}`),
      this.retryConfig
    );
  }

  /**
   * Bulk delete contacts
   */
  public async bulkDelete(ids: string[]): Promise<{ success: true }> {
    return withRetry(
      () => this.http.delete('/api/v1/contacts/bulk', { data: { ids } }),
      this.retryConfig
    );
  }

  /**
   * Bulk create/update contacts
   */
  public async bulkUpsert(contacts: CreateContactRequest[]): Promise<{ success: true; data: BulkContactResult }> {
    return withRetry(
      () => this.http.post('/api/v1/contacts/bulk', { contacts }),
      this.retryConfig
    );
  }

  /**
   * Import contacts from CSV
   */
  public async importCsv(file: File, listId?: string): Promise<{ success: true; data: ContactImportJob }> {
    const formData = new FormData();
    formData.append('file', file);
    if (listId) {
      formData.append('listId', listId);
    }
    
    return withRetry(
      () => this.http.post('/api/v1/contacts/import', formData),
      this.retryConfig
    );
  }

  /**
   * Remove all tags from a contact
   */
  public async removeTags(id: string): Promise<{ success: true }> {
    return withRetry(
      () => this.http.delete(`/api/v1/contacts/${id}/tags`),
      this.retryConfig
    );
  }

  /**
   * Remove a specific tag from a contact
   */
  public async removeTag(id: string, tag: string): Promise<{ success: true }> {
    return withRetry(
      () => this.http.delete(`/api/v1/contacts/${id}/tags/${encodeURIComponent(tag)}`),
      this.retryConfig
    );
  }

  // ========== Contact Lists ==========

  /**
   * List contact lists
   */
  public async listLists(options?: {
    page?: number;
    limit?: number;
  }): Promise<PaginatedResponse<ContactList>> {
    const params = new URLSearchParams();
    if (options?.page) params.set('page', options.page.toString());
    if (options?.limit) params.set('limit', options.limit.toString());
    
    return withRetry(
      () => this.http.get(`/api/v1/contacts/lists?${params.toString()}`),
      this.retryConfig
    );
  }

  /**
   * Get a contact list
   */
  public async getList(id: string): Promise<{ success: true; data: ContactList }> {
    return withRetry(
      () => this.http.get(`/api/v1/contacts/lists/${id}`),
      this.retryConfig
    );
  }

  /**
   * Create a contact list
   */
  public async createList(request: CreateContactListRequest): Promise<{ success: true; data: ContactList }> {
    return withRetry(
      () => this.http.post('/api/v1/contacts/lists', request),
      this.retryConfig
    );
  }

  /**
   * Update a contact list
   */
  public async updateList(id: string, request: UpdateContactListRequest): Promise<{ success: true; data: ContactList }> {
    return withRetry(
      () => this.http.patch(`/api/v1/contacts/lists/${id}`, request),
      this.retryConfig
    );
  }

  /**
   * Delete a contact list
   */
  public async deleteList(id: string): Promise<{ success: true }> {
    return withRetry(
      () => this.http.delete(`/api/v1/contacts/lists/${id}`),
      this.retryConfig
    );
  }

  // ========== List Membership ==========

  /**
   * Get list members
   */
  public async getListMembers(id: string, options?: {
    page?: number;
    limit?: number;
  }): Promise<PaginatedResponse<Contact>> {
    const params = new URLSearchParams();
    if (options?.page) params.set('page', options.page.toString());
    if (options?.limit) params.set('limit', options.limit.toString());
    
    return withRetry(
      () => this.http.get(`/api/v1/contacts/lists/${id}/members?${params.toString()}`),
      this.retryConfig
    );
  }

  /**
   * Add contacts to a list
   */
  public async addToList(id: string, contactIds: string[]): Promise<{ success: true }> {
    return withRetry(
      () => this.http.post(`/api/v1/contacts/lists/${id}/members`, { contactIds }),
      this.retryConfig
    );
  }

  /**
   * Remove contacts from a list
   */
  public async removeFromList(id: string, contactIds: string[]): Promise<{ success: true }> {
    return withRetry(
      () => this.http.delete(`/api/v1/contacts/lists/${id}/members`, { data: { contactIds } }),
      this.retryConfig
    );
  }
}

File: python/northrelay/resources/contacts.py
"""Resource module"""
from __future__ import annotations

"""Contacts resource - Contact and contact list management"""

from typing import Any, Optional
from northrelay.utils.http import HttpClient
from northrelay.utils.retry import with_retry, RetryConfig
from northrelay.types import (
    Contact,
    CreateContactRequest,
    ContactList,
    PaginatedResponse,
)


class ContactsResource:
    """Contact and contact list management"""

    def __init__(self, http: HttpClient, retry_config: RetryConfig):
        self._http = http
        self._retry_config = retry_config

    # ========== Contacts ==========

    async def list(
        self,
        *,
        page: int = 1,
        limit: int = 20,
        search: Optional[str] = None,
        list_id: Optional[str] = None,
        tags: Optional[str] = None,
    ) -> PaginatedResponse:
        """List contacts"""
        params = {"page": page, "limit": limit}
        if search:
            params["search"] = search
        if list_id:
            params["listId"] = list_id
        if tags:
            params["tags"] = tags

        response = await with_retry(
            lambda: self._http.get("/api/v1/contacts", params=params)
        )
        return PaginatedResponse(**response)

    async def create(self, request: CreateContactRequest) -> Contact:
        """Create a new contact"""
        async def _create() -> dict[str, Any]:
            payload = request.model_dump(by_alias=True, exclude_none=True)
            result = await self._http.post("/api/v1/contacts", json=payload)
            return result["data"]

        response = await with_retry(_create)
        return Contact(**response)

    async def delete(self, id: str) -> dict[str, Any]:
        """Delete a contact"""
        return await with_retry(
            lambda: self._http.delete(f"/api/v1/contacts/{id}")
        )

    async def bulk_delete(self, ids: list[str]) -> dict[str, Any]:
        """Bulk delete contacts"""
        return await with_retry(
            lambda: self._http.post("/api/v1/contacts", json={"ids": ids})
        )

    async def bulk_upsert(self, contacts: list[CreateContactRequest]) -> dict[str, Any]:
        """Bulk create/update contacts"""
        payload = {
            "contacts": [c.model_dump(by_alias=True, exclude_none=True) for c in contacts]
        }
        return await with_retry(
            lambda: self._http.post("/api/v1/contacts/bulk", json=payload)
        )

    async def import_csv(
        self, file_path: str, list_id: Optional[str] = None
    ) -> dict[str, Any]:
        """
        Import contacts from CSV
        
        Args:
            file_path: Path to CSV file
            list_id: Optional list ID to add contacts to
        """
        # Note: Python SDK uses file path, not File object like TS
        with open(file_path, "rb") as f:
            files = {"file": f}
            data = {"listId": list_id} if list_id else {}
            
            return await with_retry(
                lambda: self._http.post(
                    "/api/v1/contacts/import",
                    data=data,
                    files=files,
                )
            )

    async def remove_tags(self, id: str) -> dict[str, Any]:
        """Remove all tags from a contact"""
        return await with_retry(
            lambda: self._http.delete(f"/api/v1/contacts/{id}/tags")
        )

    async def remove_tag(self, id: str, tag: str) -> dict[str, Any]:
        """Remove a specific tag from a contact"""
        return await with_retry(
            lambda: self._http.delete(f"/api/v1/contacts/{id}/tags/{tag}")
        )

    # ========== Contact Lists ==========

    async def list_lists(
        self, *, page: int = 1, limit: int = 20
    ) -> PaginatedResponse:
        """List contact lists"""
        params = {"page": page, "limit": limit}
        response = await with_retry(
            lambda: self._http.get("/api/v1/contacts/lists", params=params)
        )
        return PaginatedResponse(**response)

    async def get_list(self, id: str) -> ContactList:
        """Get a contact list"""
        async def _get() -> dict[str, Any]:
            result = await self._http.get(f"/api/v1/contacts/lists/{id}")
            return result["data"]

        response = await with_retry(_get)
        return ContactList(**response)

    async def create_list(self, name: str, description: Optional[str] = None) -> ContactList:
        """Create a contact list"""
        async def _create() -> dict[str, Any]:
            result = await self._http.post(
                "/api/v1/contacts/lists",
                json={"name": name, "description": description},
            )
            return result["data"]

        response = await with_retry(_create)
        return ContactList(**response)

    async def update_list(
        self, id: str, name: Optional[str] = None, description: Optional[str] = None
    ) -> ContactList:
        """Update a contact list"""
        async def _update() -> dict[str, Any]:
            payload = {}
            if name:
                payload["name"] = name
            if description is not None:
                payload["description"] = description
            
            result = await self._http.post(
                f"/api/v1/contacts/lists/{id}",
                json=payload,
            )
            return result["data"]

        response = await with_retry(_update)
        return ContactList(**response)

    async def delete_list(self, id: str) -> dict[str, Any]:
        """Delete a contact list"""
        return await with_retry(
            lambda: self._http.delete(f"/api/v1/contacts/lists/{id}")
        )

    # ========== List Membership ==========

    async def get_list_members(
        self, id: str, *, page: int = 1, limit: int = 20
    ) -> PaginatedResponse:
        """Get list members"""
        params = {"page": page, "limit": limit}
        response = await with_retry(
            lambda: self._http.get(f"/api/v1/contacts/lists/{id}/members", params=params)
        )
        return PaginatedResponse(**response)

    async def add_to_list(self, id: str, contact_ids: list[str]) -> dict[str, Any]:
        """Add contacts to a list"""
        return await with_retry(
            lambda: self._http.post(
                f"/api/v1/contacts/lists/{id}/members",
                json={"contactIds": contact_ids},
            )
        )

    async def remove_from_list(self, id: str, contact_ids: list[str]) -> dict[str, Any]:
        """Remove contacts from a list"""
        return await with_retry(
            lambda: self._http.delete(
                f"/api/v1/contacts/lists/{id}/members",
                json={"contactIds": contact_ids},
            )
        )
